### DISCUSSION TITLE
0.8.0rc1 MSVC 32-bit builds fail on Falcon with AVX

### DISCUSSION THREAD

**Discussion by blackbird1999** [open-quantum-safe/liboqs#1487] [General]
44>D:\CDK8.1\liboqs\src\sig\falcon\pqclean_falcon-512_avx2\sign.c(1033): error C2440: 'function': cannot convert fr
       om 'int' to '__m128i' [D:\CDK8.1\liboqs\buildmt\src\sig\falcon\falcon_512_avx2.vcxproj]
    41>D:\CDK8.1\liboqs\src\sig\falcon\pqclean_falcon-1024_avx2\sign.c(1033): error C2440: 'function': cannot convert f
       rom 'int' to '__m128i' [D:\CDK8.1\liboqs\buildmt\src\sig\falcon\falcon_1024_avx2.vcxproj]

It appears that the AVX2 code no longer matches that found here https://falcon-sign.info/falcon-round3.zip. In particular, it no longer has the #ifdef for x64/x86 that allows it to compile. 

I'm not sure of the steps to report such things. Thanks for your help!

Sincerely,
Jonathan

**Comment by dstebila** [comment]
I'm a bit confused. Are you trying to build it with AVX / AVX2 in a 32-bit build?  If that's the case, I think that should fail, as AVX isn't available on 32-bit platforms.  Or is the problem that you're trying to make a 32-bit build but AVX code is set to be built when it shouldn't be?  

That being said, to my knowledge we've never tested 32-bit builds on Windows and don't have build infrastructure to debug or test this.

**Reply by blackbird1999** [reply]
Thanks for taking the time to reply. I appreciate it! Yes, it's the AVX2 implementation and it does work if the code from the original Falcon source is used. 

Here's the original code from falcon-sign.info's sign.c in their AVX2 implementation:
`   /*
     * We must look at the low bits for all values for which the
     * high bits are an "equal" match; values 8-18 all have the
     * same high bits (0).
     * On 32-bit systems, 'lo' really is two registers, requiring
     * some extra code.
     */
 #if defined(__x86_64__) || defined(_M_X64)
	xlo = _mm256_broadcastq_epi64(_mm_cvtsi64_si128(*(int64_t *)&lo));
#else
	{
		uint32_t e0, e1;
		int32_t f0, f1;

		e0 = (uint32_t)lo;
		e1 = (uint32_t)(lo >> 32);
		f0 = *(int32_t *)&e0;
		f1 = *(int32_t *)&e1;
		xlo = _mm256_set_epi32(f1, f0, f1, f0, f1, f0, f1, f0);
	}
#endif`

and the code currently in OQS:
`    /*
     * We must look at the low bits for all values for which the
     * high bits are an "equal" match; values 8-18 all have the
     * same high bits (0).
     * On 32-bit systems, 'lo' really is two registers, requiring
     * some extra code.
     */
    xlo = _mm256_broadcastq_epi64(_mm_cvtsi64_si128(*(int64_t *)&lo));
`
Note that the comment contains the same information about 32-bit systems, but, for some reason, the #if is missing and only the 64-bit code is present. If I merge the original in all seems fine. 

Sincerely,
Jonathan
